### ISSUE TITLE
HammerDb crashes in tclodbc

### ISSUE THREAD

**Issue by i25652** [TPC-Council/HammerDB#72]
Hi, 

Version 3.2 and 2.x crashes in tclodbc.dll the following scenario:

create database schema, create users, start load. Stop Sql server.  Users will stop with red cross marks.
Press Stop in HammerDb GUI.
The GUI will crash.

**Comment by sm-shaw** [comment]
Hi,

I have not been able to reproduce.  I start the schema creation and then stop SQL Server from SMSS. HammerDB reports the error messages below. Then press stop and the GUI does not crash. It is then possible to restart SQL Server, delete the schema and then repeat the build as normal. 

Also please provide reasons for why it would be desired to stop SQL Server while it is executing a schema build?

```
Error in Virtual User 1: S1000 596 {[Microsoft][ODBC Driver 13 for SQL Server][SQL Server]Cannot continue the execution because the session is in the kill state.}
...
Error in Virtual User 5: 08S01 233 {[Microsoft][ODBC Driver 13 for SQL Server]Shared Memory Provider: No process is on the other end of the pipe.}
Error in Virtual User 7: 08S01 233 {[Microsoft][ODBC Driver 13 for SQL Server]Shared Memory Provider: No process is on the other end of the pipe.}
...
```

**Comment by i25652** [comment]
Thank you for prompt reply!

Sorry, for confusion,  it doesn't crash on schema creation, but on actual load.  I mean after you create the schema and start transactions.  It doesn't happen with 1 virtual user, but with 10 for sure.

If those 10 users started but never connected to SQL Server (sql server hasn't been started) then they stopped/destroyed well.
If they started the load and then SQL Server stopped, then users turned to error state (as expected) and if then destroyed -> crash happens.

We use HammerDb to test application that works closely with SQL Server. The load crashes the application and then it crashes SQL Server. Right now we restart HammerDb manually over and over.

Thank you for your attention.

**Comment by sm-shaw** [comment]
Hi,
So I can now reproduce the issue.   It also occurs in the command line version so not dependent on the GUI and does not happen when keying and thinking time is enabled, so transactions are not running at a high frequency when SQL Server is disconnected. 

As follow up questions:
1. Can you please explain why you need to stop SQL Server during a test?  This does appear to be an 'edge case'. I am looking to understand the benefits of stopping the database during the test while the users are running ie If you know you want to close the database why not stop the workload first?

2. You say " load crashes the application and then it crashes SQL Server"  - do you mean that another application crashes and also SQL Server itself crashes?  So there are multiple crashes when SQL Server is stopped? 

Further understanding is needed on the circumstances on what is being tested, the overall impact and then whether a fix is warranted.

**Comment by sm-shaw** [comment]
Further investigation suggests that the issue is in the Tclodbc/driver layer as a trace shows that SQLFreeStmt with an option of SQL_DROP is entered but does not exit.  SQLFreeStmt is deprecated and maps to SQLFreeHandle.  Early testing with tdbc::odbc that uses ODBC v3 instead of ODBC v1 (TclODBC) does not call SQLFreeStmt with SQL_DROP and does not have the same issue, closing the threads after the database has been forcibly closed does clean up correctly. Therefore a potential option to resolve this issue may be migration to tdbc::odbc however discovery has shown other unrelated issues when using this interface. Current status is investigation into tdbc::odbc to determine whether this is a viable alternative.

**Comment by i25652** [comment]
Steve,

Thank you for your attention.
Yes, I think some memory freed twice and the destructor in tclodbc.dll  calls SQLFreeHandle with the garbage value. I was going to fix it by myself, but didn't found the latest source quickly and my enthusiasm was over.
Finally, I ended up with calling Hammerdb executable in the loop.
If noone reported this before, then the importance is very low.

**Comment by sm-shaw** [comment]
Many thanks for the update. I will leave this open as it is an issue with tclodbc but non-critical due to the edge-case of where the issue occurs. Also unfortunately at present tdbc::odbc is not a viable alternative.  It will be further investigated when updating the binaries.

**Comment by sm-shaw** [comment]
Good news, the issue in tdbc::odbc has been fixed and this updated version resolves this issue. When the database is shut down the following error is shown but then closing the virtual users does not result in a crash. 
```
Error in Virtual User 1: New Order : [Microsoft][ODBC Driver 13 for SQL Server]Communication link failure
[Microsoft][ODBC Driver 13 for SQL Server]The connection is broken and recovery is not possible. The client driver attempted to recover the connection one or more times and all attempts failed. Increase the value of ConnectRetryCount to increase the number of recovery attempts.
(executing the statement)
```
Plan is to complete the testing with tdbc::odbc and if successful move to that to fix this issue.

**Comment by i25652** [comment]
Steven, 

Thank you for fixing this.
I think I understand your feeling :) The rational part of you thinks "nobody seen this bug for years, let's leave it". Irrational part thinks "I love what I do and love my creatures. Let's do it perfectly."

**Comment by sm-shaw** [comment]
Absolutely.  If there is a better way to do something then it is worth pursuing even if the path is not easy at the outset.  If there is a bug then it should be fixed regardless of how often it occurs.  The TCL team have been very helpful in providing a version of tdbc::odbc that fixes the issue I found there.  Also I am working on a feature that enables scaling up to much larger session counts with keying and thinking time - this is committed for Oracle on the branch here https://github.com/sm-shaw/HammerDB/tree/44 and the plan was to do SQL Server next. However there was no point in doing this if the interface was known to have a problem that could be resolved by moving to the newer one. Therefore it made a lot of sense and saves work to do this first and write the session scaling workload with the new interface rather than the old one.

**Comment by sm-shaw** [comment]
Issue is fixed in release v3.3 https://github.com/TPC-Council/HammerDB/releases/tag/v3.3.  Note that RAISEERROR must be set to true in the driver script especially with the timed driver script with output suppressed to report the error detected. 
Having done this HammerDB will detect the disconnection as a result of the shutdown of the database and subsequently closing the virtual users will not crash the GUI.
